packages/react-components/src/components/Select/Select.tsx

```diff
@@ -0,0 +1,371 @@
+      listRef.current.scrollTop = 0;
+      hideSelectBody();
+    }
+  };
+
+  const onBodyOpen = () => {
+    document.addEventListener('click', onDocumentClick);
+
+    if (search && searchInputRef.current) {
+      timerId = setTimeout(() => {
+        searchInputRef?.current?.focus();
+      }, 150);
+    }
+  };
+
+  const onBodyClose = () => {
+    document.removeEventListener('click', onDocumentClick);
+
+    if (timerId) {
+      clearTimeout(timerId);
+    }
+  };
+
+  React.useEffect(() => {
+    if (isOpen) {
+      onDropdownToggle(true);
+      onBodyOpen();
+    }
+  }, []);
+
+  React.useEffect(() => {
+    if (getIsOpen()) {
+      return onBodyOpen();
+    }
+
+    if (!getIsOpen()) {
+      onBodyClose();
+    }
+  }, [getIsOpen]);
+
+  const filterItem = (item: any): any => {
+    if (searchPhrase) {
+      if (typeof searchProperty === 'string') {
+        if (!(searchProperty in item.props)) {
+          return false;
+        }
+
+        return item.props[searchProperty]
+          .toLocaleLowerCase()
+          .includes(searchPhrase.toLocaleLowerCase());
+      }
+
+      if (Array.isArray(searchProperty) && searchProperty.length > 0) {
+        const validSearchProperties = searchProperty.filter(
+          (p) => item.props[p]
+        );
+
+        if (validSearchProperties.length === 0) {
+          return false;
+        }
+        return validSearchProperties.some((p) =>
+          item.props[p]
+            .toLocaleLowerCase()
+            .includes(searchPhrase.toLocaleLowerCase())
+        );
+      }
+    }
+
+    return true;
+  };
+
+  const onSearchChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setSearchPhrase(e.target.value);
+
+    const filteredItems = items.filter(
+      (v) => filterItem(v) && selected !== v.key
+    );
+    const focusedItemKey =
+      filteredItems.length > 0 ? filteredItems[0].key : null;
+
+    setFocusedItemKey(focusedItemKey);
+
+    if (onSearchPhraseChange) {
+      onSearchPhraseChange(searchPhrase);
+    }
+  };
+
+  const showSelectBody = () => {
+    setIsOpen(true);
+    setSearchPhrase('');
+    onDropdownToggle(true);
+  };
+
+  const hideSelectBody = () => {
+    setIsOpen(false);
+    setFocusedItemKey(items[0] ? items[0].key : null);
+    setSearchPhrase('');
+    onDropdownToggle(false);
+
+    if (headRef.current) {
+      headRef.current.focus();
+    }
+  };
+
+  const onSelectHeadClick = (e: any) => {
+    e.preventDefault();
+
+    if (disabled) {
+      return;
+    }
+
+    if (clearButtonRef.current && clearButtonRef.current.contains(e.target)) {
+      return;
+    }
+
+    if (!getIsOpen()) {
+      showSelectBody();
+    } else {
+      hideSelectBody();
+    }
+  };
+
+  const onArrowPress = (e: KeyboardEvent) => {
+    if (e.key === EventKeys.ArrowDown || e.key === EventKeys.ArrowUp) {
+      e.preventDefault();
+      showSelectBody();
+    }
+  };
+
+  const onSelectHeadFocus = () => {
+    setIsFocused(true);
+
+    if (!isOpen) {
```

Comment:
Why in some cases we check `isOpen` which is an internal state, and sometimes we check `getIsOpened` ?